{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "0b47a2ee-5461-4d0a-a77e-4cf29cc976b7",
      "metadata": {},
      "source": [
        "# LQB385 Bioinformatics Workshop 3 - Analysis Notebook\n",
        "(Write scenario here - I suggest that this notebook was provided to the student by their bioinformatician; they now need to modify the code in here to answer questions from the PI   *-Mitchell*)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8c2836c5-b93c-4e3b-9d51-0a24b7ed04cd",
      "metadata": {
        "scrolled": true,
        "slideshow": {
          "slide_type": ""
        },
        "tags": [],
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Project startup (always upload analysed_GSE63310.RData and run this cell when notebook is first opened)\n",
        "# NOTE TO USER ---------------------------\n",
        "# Do not modify code in this cell!\n",
        "# Else it won't work!\n",
        "\n",
        "# Install packages\n",
        "install.packages(\n",
        "    packages = c(\n",
        "        \"ggpubr\",\n",
        "        \"ggsignif\",\n",
        "        \"limma\",\n",
        "        \"magrittr\",\n",
        "        \"metan\",\n",
        "        \"rlang\",\n",
        "        \"tibble\",\n",
        "        \"tidyverse\"\n",
        "    ),\n",
        "    repos = c(\n",
        "        \"https://repo.r-wasm.org\",\n",
        "        \"https://bioc.r-universe.dev\",\n",
        "        \"https://nepem-ufsc.r-universe.dev\",\n",
        "        \"https://rstudio.r-universe.dev\",\n",
        "        \"https://r-lib.r-universe.dev\",\n",
        "        \"https://tidyverse.r-universe.dev\"\n",
        "    )\n",
        ")\n",
        "\n",
        "# Import packages\n",
        "library(\"ggpubr\")\n",
        "library(\"ggsignif\")\n",
        "library(\"limma\")\n",
        "library(\"magrittr\")\n",
        "library(\"metan\")\n",
        "library(\"rlang\")\n",
        "library(\"tibble\")\n",
        "library(\"tidyverse\")\n",
        "\n",
        "# Define functions used in this notebook\n",
        "symbol_to_entrez <- function(EList, symbols) {\n",
        "    # This function accepts an EList containing voom-normalised logCPM data; a vector of characters representing genes of interest\n",
        "    # and outputs the corresponding indices.\n",
        "    # With error catching!\n",
        "    indices <- which(EList$genes$SYMBOL %in% symbols)\n",
        "\n",
        "    ## Check: Error if there are fewer ENTREZ IDs than gene symbols (impossible)\n",
        "    if (length(indices) < length(symbols)) {\n",
        "        stop(\n",
        "            \"There are fewer gene IDs than there should be. Check your gene symbols!\"\n",
        "        )\n",
        "    }\n",
        "\n",
        "    return(indices)\n",
        "}\n",
        "\n",
        "voom_to_long <- function(EList, symbols) {\n",
        "    # This function accepts an EList containing voom-normalised logCPM data;\n",
        "    # a character representing gene of interest\n",
        "    # and outputs a dataframe ready for tidyverse data analysis\n",
        "\n",
        "    # Get indices of selected genes\n",
        "    indices <- symbol_to_entrez(EList, symbols)\n",
        "\n",
        "    # Filter EList for selected indices\n",
        "    # Rename rownames into gene symbols\n",
        "    # Then transpose\n",
        "    # Then convert into dataframe\n",
        "    # Build 2 cases: if there is one gene, treat extracted data as a vector.\n",
        "    # If there are multiple, treat extracted data like a df\n",
        "    if (length(indices) == 1) {\n",
        "        # Add column for sample_id\n",
        "        lcpm_selected <- EList$E[indices, ]\n",
        "        lcpm_selected <- tibble(\n",
        "            sample_id = names(lcpm_selected),\n",
        "            logCPM = lcpm_selected\n",
        "        )\n",
        "\n",
        "        # Get samplesheet\n",
        "        # Append expression data\n",
        "        # Make tidy\n",
        "        lcpm_output <- EList$targets %>%\n",
        "            rownames_to_column(\"sample_id\") %>%\n",
        "            left_join(\n",
        "                lcpm_selected,\n",
        "                by = join_by(\"sample_id\" == \"sample_id\"),\n",
        "                unmatched = \"error\"\n",
        "            )\n",
        "\n",
        "        # Return data\n",
        "        return(lcpm_output)\n",
        "    } else {\n",
        "        # Add column for sample_id\n",
        "        lcpm_selected <- EList$E[indices, ] %>%\n",
        "            set_rownames(symbols) %>%\n",
        "            t() %>%\n",
        "            as.data.frame() %>%\n",
        "            rownames_to_column(\"sample_id\")\n",
        "\n",
        "        # Get samplesheet\n",
        "        # Append expression data\n",
        "        # Make tidy\n",
        "        lcpm_output <- EList$targets %>%\n",
        "            rownames_to_column(\"sample_id\") %>%\n",
        "            left_join(\n",
        "                lcpm_selected,\n",
        "                by = join_by(\"sample_id\" == \"sample_id\"),\n",
        "                unmatched = \"error\"\n",
        "            ) %>%\n",
        "            pivot_longer(\n",
        "                cols = symbols,\n",
        "                names_to = \"SYMBOL\",\n",
        "                values_to = \"logCPM\"\n",
        "            )\n",
        "\n",
        "        # Return data\n",
        "        return(lcpm_output)\n",
        "    }\n",
        "}\n",
        "\n",
        "get_signif_gse <- function(MArrayLM, symbol, contrasts) {\n",
        "    # This function accepts a contrasted MArrayLM,\n",
        "    # a character representing the gene of interest\n",
        "    # a vector of contrasts to extract data from\n",
        "    # and outputs a vector of adjusted p values.\n",
        "\n",
        "    out <- map(contrasts, \\(contrast) {\n",
        "        topTable(\n",
        "            fit = MArrayLM,\n",
        "            coef = contrast,\n",
        "            number = Inf,\n",
        "            sort.by = \"none\"\n",
        "        ) %>%\n",
        "            filter(SYMBOL == symbol) %>%\n",
        "            .$adj.P.Val\n",
        "    }) %>%\n",
        "        unlist() %>%\n",
        "        stars_pval()\n",
        "\n",
        "    return(out)\n",
        "}\n",
        "\n",
        "# Load data\n",
        "load(\"./analysed_GSE63310.Rdata\")\n",
        "\n",
        "# Show list of imported packages (debug only)\n",
        "sessionInfo()\n",
        "# Show list of data (debug only)\n",
        "# ls()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "fc5ae67b-57b6-4c08-86c8-04a279e7bc0c",
      "metadata": {},
      "source": [
        "# Analysis of genes of interest"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "34ec5a43-eaa5-4cd5-a0db-52ee58dbfac7",
      "metadata": {},
      "source": [
        "This section includes code cells for the following analyses:\n",
        "- Gene expression histogram\n",
        "  - Visualisation of expression of one gene.\n",
        "- Gene expression boxplots\n",
        "  - Visualisation of expression of one gene across experimental conditions.\n",
        "- Correlation scatterplot\n",
        "  - Visualisation of correlation between two selected genes.\n",
        "- Correlation heatmap\n",
        "  - Visualisation of correlation between selected genes.\n",
        "\n",
        "All these analyses use logCPM data."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "11c595df-b478-4e08-97ca-c041cdd05ab4",
      "metadata": {},
      "source": [
        "## Gene expression histogram\n",
        "The following code produces a histogram of gene expression for the selected gene(s).\n",
        "\n",
        "Use `?ggplot` and `?geom_histogram` to understand the code further."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c0e3c0a6-58a7-43a4-a900-dd7ade42fe42",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Set gene(s) to be plotted\n",
        "## Remember that Mus Musculus gene names are in Sentence Case\n",
        "## e.g. Sulf1 Acta1\n",
        "current_gene <- \"Sulf1\"\n",
        "\n",
        "# Extract gene expression data\n",
        "lcpm_histogram <- voom_to_long(GSE_logCPM, current_gene)\n",
        "\n",
        "# Construct histogram\n",
        "plot_histogram <- ggplot(\n",
        "    lcpm_histogram,\n",
        "    aes(x = logCPM)\n",
        ") +\n",
        "    geom_histogram() +\n",
        "    theme_minimal() +\n",
        "    ggtitle(\n",
        "      label = current_gene |>\n",
        "        str_replace_all(\"_\", \" \") |>\n",
        "        str_wrap(width = 20, whitespace_only = TRUE)\n",
        "    )\n",
        "\n",
        "# Draw histogram\n",
        "plot(plot_histogram)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0e210a9c-c9ae-468c-86ff-1b10c3f735b0",
      "metadata": {},
      "source": [
        "## Gene expression boxplots\n",
        "The following code produces a boxplot of gene expression changes for the selected gene.\n",
        "\n",
        "Statistical analysis was performed using the limma moderated t-test. Adjusted p values: `<0.05: *. <0.01: **. <0.001: ***. <0.0001: ****`\n",
        "\n",
        "Use `?ggplot` and `?geom_boxplot` to understand the code further."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "51369ec0-8c2f-4c68-8983-5110eba962d9",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "scrolled": true,
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Set gene(s) to be plotted\n",
        "current_gene <- \"Sulf1\"\n",
        "if (length(current_gene) > 1) stop(\"This plot can only handle 1 gene\")\n",
        "\n",
        "# Extract gene expression data\n",
        "lcpm_boxplot <- voom_to_long(GSE_logCPM, current_gene)\n",
        "\n",
        "# Construct histogram\n",
        "plot_box <- ggplot(\n",
        "    lcpm_boxplot,\n",
        "    aes(x = group,y = logCPM)\n",
        ") +\n",
        "    geom_boxplot(aes(colour = group), outliers = FALSE) +\n",
        "    geom_jitter(\n",
        "        size = 1.2,\n",
        "        stroke = 0.5,\n",
        "        alpha = 0.5\n",
        "    ) + \n",
        "    geom_signif(\n",
        "        comparisons = list(\n",
        "            c(\"Basal\", \"LP\"),\n",
        "            c(\"Basal\", \"ML\"),\n",
        "            c(\"LP\", \"ML\")\n",
        "        ),\n",
        "        annotation = get_signif_gse(\n",
        "            GSE_fitted,\n",
        "            current_gene,\n",
        "            c(\"lp_vs_basal\", \"ml_vs_basal\", \"ml_vs_lp\")\n",
        "        ),\n",
        "        y_position = rep(max(lcpm_boxplot$logCPM) * 1.1, 3),\n",
        "        textsize = 3,\n",
        "        step_increase = 0.15\n",
        "    ) +\n",
        "    scale_y_continuous(expand = expansion(0, 0.4)) +\n",
        "    theme_minimal() +\n",
        "    ggtitle(\n",
        "      label = current_gene |>\n",
        "        str_replace_all(\"_\", \" \") |>\n",
        "        str_wrap(width = 20, whitespace_only = TRUE)\n",
        "    )\n",
        "\n",
        "# Draw histogram\n",
        "plot(plot_box)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8afe7b5b-5695-4308-b00f-19f2545ece25",
      "metadata": {},
      "source": [
        "## Gene-gene correlation scatterplot\n",
        "The following code produces a scatterplot of gene expression for the selected 2 genes.\n",
        "\n",
        "Statistical analysis was performed using Pearson's correlation."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0db0dafd-6a0c-4bee-b897-276196d0792e",
      "metadata": {
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Set gene(s) to be plotted\n",
        "current_gene <- c(\"Gpc1\",\"Sulf1\")\n",
        "\n",
        "if (length(current_gene) != 2) stop(\"This plot can only handle 2 genes\")\n",
        "\n",
        "# Extract gene expression data\n",
        "lcpm_scatter <- voom_to_long(GSE_logCPM, current_gene)\n",
        "lcpm_scatter <- pivot_wider(lcpm_scatter, names_from = SYMBOL, values_from = logCPM)\n",
        "\n",
        "# Construct plot\n",
        "## Define x and y axis based on selected genes\n",
        "scatter_aesthetics <- current_gene %>% \n",
        "    set_names(c(\"x\", \"y\")) %>%\n",
        "    map(sym) %>%\n",
        "    as_quosures(env = .GlobalEnv)\n",
        "\n",
        "## Draw plot with defined aesthetics\n",
        "plot_scatter <- ggplot(\n",
        "    lcpm_scatter,\n",
        "    aes(!!!scatter_aesthetics)\n",
        ") +\n",
        "    geom_point(\n",
        "        aes(color = group),\n",
        "        size = 1.2,\n",
        "        stroke = 0.5,\n",
        "        alpha = 0.8\n",
        "    ) +\n",
        "    geom_smooth(\n",
        "        colour = \"grey80\",\n",
        "        method = \"lm\",\n",
        "        se = FALSE\n",
        "    ) +\n",
        "    stat_cor(\n",
        "        aes(label = paste(..r.label.., ..p.label.., sep = \"~`,`~\")),\n",
        "        r.accuracy = 0.01,\n",
        "        alpha = 1\n",
        "    ) +\n",
        "    theme_minimal()\n",
        "\n",
        "# Draw plot\n",
        "plot(plot_scatter)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f40ccf31-7740-439c-b3e7-b196f0a74dbb",
      "metadata": {},
      "source": [
        "## Gene-gene correlation heatmap\n",
        "The following code produces a heatmap of gene expression correlation for the selected genes.\n",
        "\n",
        "Statistical analysis was performed using Pearson's correlation."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "id": "7d6c78a9-dd25-44eb-90a7-1b9c50728d18",
      "metadata": {
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Set gene(s) to be plotted\n",
        "current_gene <- c(\"Gpc1\",\"Sulf1\", \"Clu\", \"Basp1\", \"Cd82\")\n",
        "\n",
        "if (length(current_gene) < 2) stop(\"This plot needs at least 2 genes\")\n",
        "\n",
        "# Extract gene expression data\n",
        "\n",
        "# Construct plot\n",
        "\n",
        "# Draw plot\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2f7a01a7-384e-421f-aa98-211cc0998062",
      "metadata": {},
      "source": [
        "# Advanced details"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "220b41bd-e8e4-4a0e-a471-568703a4d729",
      "metadata": {},
      "source": [
        "## R syntax - variable assignment\n",
        "\n",
        "In some cases throughout this workshop, you will need to assign one or many genes to a variable.\n",
        "\n",
        "To assign one gene, use: `x <- \"Gene\"`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8c6dcd56-239f-43dc-93d6-0b2c8f0abc08",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Demonstration: assignment of one character to one variable\n",
        "x <- \"Gene\"\n",
        "\n",
        "# Show content of x\n",
        "print(x)\n",
        "\n",
        "# Clean up (remove x)\n",
        "rm(x)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "db9c3faf-da4b-47b4-9236-4c93deb2f947",
      "metadata": {},
      "source": [
        "To assign multiple genes (a vector of gene symbols), use: `x <- c(\"Gene1\", \"Gene2\", \"Gene3\", \"Gene4\")`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "2b6c6da5-9348-4822-b1bd-66a5502f3552",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Demonstration: assignment of a vector to one variable\n",
        "x <- c(\"Gene1\", \"Gene2\", \"Gene3\", \"Gene4\")\n",
        "\n",
        "# Longer vectors can span over multiple lines\n",
        "x <- c(\n",
        "    \"Gene1\",\n",
        "    \"Gene2\",\n",
        "    \"Gene3\",\n",
        "    \"Gene4\"\n",
        ")\n",
        "\n",
        "# Show content of x\n",
        "print(x)\n",
        "\n",
        "# Clean up (remove x)\n",
        "rm(x)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6bc495c4-dbe6-4116-ad52-f61713f84585",
      "metadata": {},
      "source": [
        "Common errors:\n",
        "- **R is case-sensitive!** Make sure you use the correct case when referring to your genes.\n",
        "- Characters must be wrapped in \"quotation marks\"."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6568026b-7922-440a-b0c3-1603fcfb0e1d",
      "metadata": {},
      "source": [
        "## R syntax - exploring objects in this project"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a5f96dca-2946-4f09-b8af-a9c75cbe487d",
      "metadata": {},
      "source": [
        "The uploaded `analysed_GSE63310.Rdata` includes the following objects:\n",
        "- `GSE_logCPM`: Log-counts per million RNA-seq data (with accompanying gene and sample information)\n",
        "- `GSE_fitted`: Test results of RNA-seq data (with accompanying gene and sample information)\n",
        "- `GSE_camera`: Gene set analysis results of RNA-seq data\n",
        "- `GSE_pca`: Principal Component Analysis results of RNA-seq data\n",
        "\n",
        "Below are code examples for accessing these data:"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0169c3f8-af06-4df6-ad7f-56e4d240819f",
      "metadata": {},
      "source": [
        "- Accessing log-counts per million data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d8e88e6d-6925-4dda-b230-6660da65eeae",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "scrolled": true,
        "trusted": true
      },
      "outputs": [],
      "source": [
        "GSE_logCPM[[\"E\"]]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e0a87297-b24c-4bac-bcf4-d75aec7f1e33",
      "metadata": {},
      "source": [
        "- Accessing gene information"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e0a171ef-4d20-4f72-a015-d02d9930034c",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "trusted": true
      },
      "outputs": [],
      "source": [
        "GSE_logCPM[[\"genes\"]]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "89768fa9-ca03-4e32-b0b1-544ba9b6a462",
      "metadata": {},
      "source": [
        "- Accessing sample information"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "be4e4dc1-4a8e-4dc8-99fc-237a20d407ed",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "trusted": true
      },
      "outputs": [],
      "source": [
        "GSE_logCPM[[\"targets\"]]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e34a73e3-6449-4097-8c36-1fc077ef9303",
      "metadata": {},
      "source": [
        "- Accessing comparisons definition (in this context, comparisons are called \"contrasts\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d48812ae-a958-4135-bd21-06de18f4799d",
      "metadata": {
        "jupyter": {
          "source_hidden": true
        },
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Accessing contrasts\n",
        "print(GSE_fitted[[\"contrasts\"]])\n",
        "\n",
        "# Accessing names of contrasts\n",
        "print(colnames(GSE_fitted[[\"contrasts\"]]))"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "30296f1b-d202-490d-9edd-2c826d111fa4",
      "metadata": {},
      "source": [
        "- Accessing gene set analysis results"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0ca4ca5f-4989-4c0c-834b-a469307889fb",
      "metadata": {
        "editable": true,
        "jupyter": {
          "source_hidden": true
        },
        "scrolled": true,
        "slideshow": {
          "slide_type": ""
        },
        "tags": [],
        "trusted": true
      },
      "outputs": [],
      "source": [
        "# Note: use GSE_camera[[\"name of contrasts\"]]\n",
        "GSE_camera[[\"ml_vs_lp\"]]"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "R (webR)",
      "language": "R",
      "name": "webR"
    },
    "language_info": {
      "name": "r"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
