{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "0b47a2ee-5461-4d0a-a77e-4cf29cc976b7",
      "metadata": {},
      "source": [
        "# LQB385 Bioinformatics Workshop 3 - Analysis Notebook\n",
        "(Write scenario here - I suggest that this notebook was provided to the student by their bioinformatician; they now need to modify the code in here to answer questions from the PI   *-Mitchell*)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "16ed0115-e975-45a3-a09a-180794715a2a",
      "metadata": {},
      "source": [
        "# Introduction (tips)\n",
        "\n",
        "**NB: just random tips atm** \n",
        "\n",
        "- Use light mode\n",
        "\n",
        "- Show list of comparisons"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "b087cb21",
      "metadata": {},
      "source": [
        "---\n",
        "format:\n",
        "  typst:\n",
        "    papersize: a4\n",
        "    margin:\n",
        "      x: 1cm\n",
        "      y: 1cm\n",
        "    columns: 1\n",
        "---"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8c2836c5-b93c-4e3b-9d51-0a24b7ed04cd",
      "metadata": {
        "editable": true,
        "slideshow": {
          "slide_type": ""
        },
        "tags": [],
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: false\n",
        "#| warning: false\n",
        "#| output: false\n",
        "\n",
        "# Project startup (always upload analysed_GSE63310.RData and run this cell when notebook is first opened)\n",
        "# NOTE TO USER ---------------------------\n",
        "# Do not modify code in this cell!\n",
        "# Else it won't work!\n",
        "\n",
        "# Install packages\n",
        "install.packages(\n",
        "    packages = c(\n",
        "        \"colorspace\",\n",
        "        \"corrplot\",\n",
        "        \"EnhancedVolcano\",\n",
        "        \"ggpubr\",\n",
        "        \"ggsignif\",\n",
        "        \"limma\",\n",
        "        \"magrittr\",\n",
        "        \"metan\",\n",
        "        \"PCAtools\",\n",
        "        \"rlang\",\n",
        "        \"tibble\",\n",
        "        \"tidyverse\"\n",
        "    ),\n",
        "    repos = c(\n",
        "        \"https://repo.r-wasm.org\",\n",
        "        \"https://bioc.r-universe.dev\",\n",
        "        \"https://cran.r-universe.dev\",\n",
        "        \"https://dmuellner.r-universe.dev\",\n",
        "        \"https://nepem-ufsc.r-universe.dev\",\n",
        "        \"https://rstudio.r-universe.dev\",\n",
        "        \"https://r-lib.r-universe.dev\",\n",
        "        \"https://tidyverse.r-universe.dev\"\n",
        "    )\n",
        ")\n",
        "\n",
        "# Import packages\n",
        "library(\"colorspace\")\n",
        "library(\"corrplot\")\n",
        "library(\"EnhancedVolcano\")\n",
        "library(\"ggpubr\")\n",
        "library(\"ggsignif\")\n",
        "library(\"limma\")\n",
        "library(\"magrittr\")\n",
        "library(\"metan\")\n",
        "library(\"PCAtools\")\n",
        "library(\"rlang\")\n",
        "library(\"tibble\")\n",
        "library(\"tidyverse\")\n",
        "\n",
        "# Define functions used in this notebook\n",
        "symbol_to_entrez <- function(EList, symbols) {\n",
        "    # This function accepts an EList containing voom-normalised logCPM data\n",
        "    # a vector of characters representing genes of interest\n",
        "    # and outputs the corresponding indices.\n",
        "    # With error catching!\n",
        "    indices <- which(EList$genes$SYMBOL %in% symbols)\n",
        "\n",
        "    ## Check: Error if there are fewer ENTREZ IDs than gene symbols (impossible)\n",
        "    if (length(indices) < length(symbols)) {\n",
        "        stop(\n",
        "            \"There are fewer gene IDs than there should be. Check your gene symbols!\"\n",
        "        )\n",
        "    }\n",
        "\n",
        "    return(indices)\n",
        "}\n",
        "\n",
        "voom_to_long <- function(EList, symbols) {\n",
        "    # This function accepts an EList containing voom-normalised logCPM data;\n",
        "    # a vector of characters representing gene of interest\n",
        "    # and outputs a dataframe ready for tidyverse data analysis\n",
        "\n",
        "    # Get indices of selected genes\n",
        "    indices <- symbol_to_entrez(EList, symbols)\n",
        "\n",
        "    # Filter EList for selected indices\n",
        "    # Rename rownames into gene symbols\n",
        "    # Then transpose\n",
        "    # Then convert into dataframe\n",
        "    # Build 2 cases: if there is one gene, treat extracted data as a vector.\n",
        "    # If there are multiple, treat extracted data like a df\n",
        "    if (length(indices) == 1) {\n",
        "        # Add column for sample_id\n",
        "        lcpm_selected <- EList$E[indices, ]\n",
        "        lcpm_selected <- tibble(\n",
        "            sample_id = names(lcpm_selected),\n",
        "            logCPM = lcpm_selected\n",
        "        )\n",
        "\n",
        "        # Get samplesheet\n",
        "        # Append expression data\n",
        "        # Make tidy\n",
        "        lcpm_output <- EList$targets %>%\n",
        "            rownames_to_column(\"sample_id\") %>%\n",
        "            left_join(\n",
        "                lcpm_selected,\n",
        "                by = join_by(\"sample_id\" == \"sample_id\"),\n",
        "                unmatched = \"error\"\n",
        "            )\n",
        "\n",
        "        # Return data\n",
        "        return(lcpm_output)\n",
        "    } else {\n",
        "        # Add column for sample_id\n",
        "        lcpm_selected <- EList$E[indices, ] %>%\n",
        "            set_rownames(symbols) %>%\n",
        "            t() %>%\n",
        "            as.data.frame() %>%\n",
        "            rownames_to_column(\"sample_id\")\n",
        "\n",
        "        # Get samplesheet\n",
        "        # Append expression data\n",
        "        # Make tidy\n",
        "        lcpm_output <- EList$targets %>%\n",
        "            rownames_to_column(\"sample_id\") %>%\n",
        "            left_join(\n",
        "                lcpm_selected,\n",
        "                by = join_by(\"sample_id\" == \"sample_id\"),\n",
        "                unmatched = \"error\"\n",
        "            ) %>%\n",
        "            pivot_longer(\n",
        "                cols = symbols,\n",
        "                names_to = \"SYMBOL\",\n",
        "                values_to = \"logCPM\"\n",
        "            )\n",
        "\n",
        "        # Return data\n",
        "        return(lcpm_output)\n",
        "    }\n",
        "}\n",
        "\n",
        "voom_to_matrix <- function(EList, symbols) {\n",
        "    # This function accepts an EList containing voom-normalised logCPM data;\n",
        "    # a vector of characters representing gene of interest\n",
        "    # and outputs a matrix for correlation heatmaps\n",
        "    \n",
        "    # Get indices of selected genes\n",
        "    indices <- symbol_to_entrez(EList, symbols)\n",
        "\n",
        "    # Filter EList for selected indices\n",
        "    # Rename rownames into gene symbols\n",
        "    lcpm_selected <- EList$E[indices, ] %>%\n",
        "        set_rownames(symbols)\n",
        "    \n",
        "    # Return data\n",
        "    return(lcpm_selected)\n",
        "}\n",
        "\n",
        "get_signif_gse <- function(MArrayLM, symbol, contrasts) {\n",
        "    # This function accepts a contrasted MArrayLM,\n",
        "    # a character representing the gene of interest\n",
        "    # a vector of contrasts to extract data from\n",
        "    # and outputs a vector of adjusted p values.\n",
        "\n",
        "    out <- map(contrasts, \\(contrast) {\n",
        "        topTable(\n",
        "            fit = MArrayLM,\n",
        "            coef = contrast,\n",
        "            number = Inf,\n",
        "            sort.by = \"none\"\n",
        "        ) %>%\n",
        "            filter(SYMBOL == symbol) %>%\n",
        "            .$adj.P.Val\n",
        "    }) %>%\n",
        "        unlist() %>%\n",
        "        stars_pval()\n",
        "\n",
        "    return(out)\n",
        "}\n",
        "\n",
        "format_gene_symbol_gse <- function(symbol) {\n",
        "    symbol |>\n",
        "        str_replace_all(\"_\", \" \") |>\n",
        "        str_wrap(width = 20, whitespace_only = TRUE)\n",
        "}\n",
        "\n",
        "# Load data\n",
        "load(\"./analysed_GSE63310.Rdata\")\n",
        "\n",
        "# Show list of imported packages (debug only)\n",
        "# sessionInfo()\n",
        "# Show list of data (debug only)\n",
        "# ls()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "fc5ae67b-57b6-4c08-86c8-04a279e7bc0c",
      "metadata": {},
      "source": [
        "# Analysis of genes of interest\n",
        "\n",
        "This section includes code cells for the following analyses:\n",
        "\n",
        "- Gene expression histogram\n",
        "  - Visualisation of expression of one gene.\n",
        "- Gene expression boxplots\n",
        "  - Visualisation of expression of one gene across experimental conditions.\n",
        "- Correlation scatterplot\n",
        "  - Visualisation of correlation between two selected genes.\n",
        "- Correlation heatmap\n",
        "  - Visualisation of correlation between selected genes.\n",
        "\n",
        "All these analyses use logCPM data."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "11c595df-b478-4e08-97ca-c041cdd05ab4",
      "metadata": {},
      "source": [
        "## Gene expression histogram\n",
        "The following code produces a histogram of gene expression for the selected gene(s).\n",
        "\n",
        "Use `?ggplot` and `?geom_histogram` to understand the code further."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c0e3c0a6-58a7-43a4-a900-dd7ade42fe42",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "\n",
        "# Set gene(s) to be plotted\n",
        "## Remember that Mus Musculus gene names are in Sentence Case\n",
        "## e.g. Sulf1 Acta1\n",
        "current_gene <- \"Sulf1\"\n",
        "\n",
        "# Extract gene expression data\n",
        "lcpm_histogram <- voom_to_long(GSE_logCPM, current_gene)\n",
        "\n",
        "# Construct histogram\n",
        "plot_histogram <- ggplot(\n",
        "    lcpm_histogram,\n",
        "    aes(x = logCPM)\n",
        ") +\n",
        "    geom_histogram() +\n",
        "    theme_minimal() +\n",
        "    ggtitle(label = format_gene_symbol_gse(current_gene))\n",
        "\n",
        "# Draw histogram\n",
        "plot(plot_histogram)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0e210a9c-c9ae-468c-86ff-1b10c3f735b0",
      "metadata": {},
      "source": [
        "## Gene expression boxplots\n",
        "The following code produces a boxplot of gene expression changes for the selected gene.\n",
        "\n",
        "Statistical analysis was performed using the limma moderated t-test. Adjusted p values: `<0.05: *. <0.01: **. <0.001: ***. <0.0001: ****`\n",
        "\n",
        "Use `?ggplot` and `?geom_boxplot` to understand the code further."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "51369ec0-8c2f-4c68-8983-5110eba962d9",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "\n",
        "# Set gene(s) to be plotted\n",
        "current_gene <- \"Sulf1\"\n",
        "if (length(current_gene) > 1) {\n",
        "    stop(\"This plot can only handle 1 gene\")\n",
        "}\n",
        "\n",
        "# Extract gene expression data\n",
        "lcpm_boxplot <- voom_to_long(GSE_logCPM, current_gene)\n",
        "\n",
        "# Construct histogram\n",
        "plot_box <- ggplot(\n",
        "    lcpm_boxplot,\n",
        "    aes(x = group,y = logCPM)\n",
        ") +\n",
        "    geom_boxplot(aes(colour = group), outliers = FALSE) +\n",
        "    geom_jitter(\n",
        "        size = 1.2,\n",
        "        stroke = 0.5,\n",
        "        alpha = 0.5\n",
        "    ) + \n",
        "    geom_signif(\n",
        "        comparisons = list(\n",
        "            c(\"Basal\", \"LP\"),\n",
        "            c(\"Basal\", \"ML\"),\n",
        "            c(\"LP\", \"ML\")\n",
        "        ),\n",
        "        annotation = get_signif_gse(\n",
        "            GSE_fitted,\n",
        "            current_gene,\n",
        "            c(\"lp_vs_basal\", \"ml_vs_basal\", \"ml_vs_lp\")\n",
        "        ),\n",
        "        y_position = rep(max(lcpm_boxplot$logCPM) * 1.1, 3),\n",
        "        textsize = 3,\n",
        "        step_increase = 0.15\n",
        "    ) +\n",
        "    scale_y_continuous(expand = expansion(0, 0.4)) +\n",
        "    theme_minimal() +\n",
        "    ggtitle(label = format_gene_symbol_gse(current_gene))\n",
        "\n",
        "# Draw histogram\n",
        "plot(plot_box)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8afe7b5b-5695-4308-b00f-19f2545ece25",
      "metadata": {},
      "source": [
        "## Gene-gene correlation scatterplot\n",
        "The following code produces a scatterplot of gene expression for the selected 2 genes.\n",
        "\n",
        "Statistical analysis was performed using Pearson's correlation."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0db0dafd-6a0c-4bee-b897-276196d0792e",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "\n",
        "# Set gene(s) to be plotted\n",
        "current_gene <- c(\"Gpc1\",\"Sulf1\")\n",
        "\n",
        "if (length(current_gene) != 2) {\n",
        "    stop(\"This plot can only handle 2 genes\")\n",
        "}\n",
        "\n",
        "# Extract gene expression data\n",
        "lcpm_scatter <- voom_to_long(GSE_logCPM, current_gene)\n",
        "lcpm_scatter <- pivot_wider(\n",
        "    lcpm_scatter,\n",
        "    names_from = SYMBOL,\n",
        "    values_from = logCPM\n",
        ")\n",
        "\n",
        "# Construct plot\n",
        "## Define x and y axis based on selected genes\n",
        "scatter_aesthetics <- current_gene |>\n",
        "    set_names(c(\"x\", \"y\")) |>\n",
        "    map(sym) |>\n",
        "    as_quosures(env = .GlobalEnv)\n",
        "\n",
        "## Draw plot with defined aesthetics\n",
        "plot_scatter <- ggplot(\n",
        "    lcpm_scatter,\n",
        "    aes(!!!scatter_aesthetics)\n",
        ") +\n",
        "    geom_point(\n",
        "        aes(color = group),\n",
        "        size = 1.2,\n",
        "        stroke = 0.5,\n",
        "        alpha = 0.8\n",
        "    ) +\n",
        "    geom_smooth(\n",
        "        colour = \"grey80\",\n",
        "        method = \"lm\",\n",
        "        se = FALSE\n",
        "    ) +\n",
        "    stat_cor(\n",
        "        aes(label = paste(\n",
        "            ..r.label..,\n",
        "            ..p.label..,\n",
        "            sep = \"~`,`~\"\n",
        "        )),\n",
        "        r.accuracy = 0.01,\n",
        "        alpha = 1\n",
        "    ) +\n",
        "    theme_minimal()\n",
        "\n",
        "# Draw plot\n",
        "plot(plot_scatter)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f40ccf31-7740-439c-b3e7-b196f0a74dbb",
      "metadata": {},
      "source": [
        "## Gene-gene correlation heatmap\n",
        "The following code produces a heatmap of gene expression correlation for the selected genes.\n",
        "\n",
        "Statistical analysis was performed using Pearson's correlation."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7d6c78a9-dd25-44eb-90a7-1b9c50728d18",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "\n",
        "# Set gene(s) to be plotted\n",
        "current_gene <- c(\"Gpc1\",\"Sulf1\", \"Clu\", \"Basp1\", \"Cd82\")\n",
        "\n",
        "if (length(current_gene) < 2) {\n",
        "    stop(\"This plot needs at least 2 genes\")\n",
        "} \n",
        "\n",
        "# Extract gene expression data\n",
        "lcpm_correlation <- voom_to_matrix(GSE_logCPM, current_gene)\n",
        "lcpm_correlation <- cor(t(lcpm_correlation))\n",
        "\n",
        "# Construct plot\n",
        "corrplot(\n",
        "    corr = lcpm_correlation,\n",
        "    method = 'color',\n",
        "    col = COL2('RdBu', 200) |> rev() |> lighten(0.2),\n",
        "    bg = \"white\",\n",
        "    addCoef.col = 'grey20',\n",
        "    order = 'AOE',\n",
        "    tl.col = \"black\",\n",
        "    number.cex = 0.8\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a64683b2-0880-4d66-84a1-3fcd1c41ad87",
      "metadata": {},
      "source": [
        "# Omics exploratory analysis\n",
        "This section includes code cells for the following analyses:\n",
        "- Principal Component Analysis (PCA) biplot:\n",
        "  - Highlight clustering of samples by expression pattern"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "92940785-781a-4b5b-9fa0-caab7400e078",
      "metadata": {},
      "source": [
        "## PCA biplot\n",
        "The following code produces a PCA biplot to demonstrate clustering of samples by variables in the dataset.\n",
        "\n",
        "The samples can be coloured using the following:\n",
        "- \"group\": Experimental group.\n",
        "- \"lane\": Sequencing batch."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "2986e01b-7ecc-4411-bb19-21a5eede1e29",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "\n",
        "# Select variable to colour samples by\n",
        "current_variable <- \"group\"\n",
        "\n",
        "# Construct PCA biplot\n",
        "plot_pca <- biplot(\n",
        "    GSE_pca,\n",
        "    x = \"PC1\",\n",
        "    y = \"PC2\",\n",
        "    colby = current_variable,\n",
        "    pointSize = 3,\n",
        "    legendPosition = \"right\"\n",
        ")\n",
        "\n",
        "# Draw plot\n",
        "plot(plot_pca)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c7dae36e-85a8-4267-a8b1-f2904ce228be",
      "metadata": {},
      "source": [
        "# Differential gene expression (DGE)\n",
        "This section includes code cells for the following analyses:\n",
        "- DGE analysis statistics:\n",
        "  - Number of genes up/downregulated in each comparison.\n",
        "- Volcano plots:\n",
        "  - Visualisation of differential expression between conditions.\n",
        "- Table of top differentially expressed genes"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "46f0cbdb-260b-41bd-a4d0-737195939ed0",
      "metadata": {},
      "source": [
        "## DGE analysis statistics\n",
        "The following code shows the number of DEGs in each comparison.\n",
        "\n",
        "A gene is considered a DEG when its `adjusted p value < 0.05`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "01f2eeb4-be6d-4ee9-aeee-c828f9943a41",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "if (length(dev.list() > 1)) dev.off()\n",
        "\n",
        "# DGE analysis summary statistics\n",
        "summary(decideTests(GSE_fitted))"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8d6185c3-836d-4aed-af1d-cc0215e16c2a",
      "metadata": {},
      "source": [
        "## Volcano plots\n",
        "The following code produces a volcano plot to demonstrate gene expression changes within each comparison.\n",
        "\n",
        "Samples are coloured as below:\n",
        "\n",
        "- **Grey**: non-significant\n",
        "- **Green**: logFC > 1\n",
        "- **Blue**: adjusted p value < 0.05\n",
        "- **Red**: adjusted p value < 0.05 and logFC > 1."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "38248717-c89d-4a85-8d52-f5760c5b23c7",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "\n",
        "# Define current contrast\n",
        "current_contrast <- \"ml_vs_basal\"\n",
        "\n",
        "# Get test statistics\n",
        "current_toptable <- topTable(\n",
        "    fit = GSE_fitted,\n",
        "    coef = current_contrast,\n",
        "    number = Inf\n",
        ")\n",
        "\n",
        "# Construct volcano plot\n",
        "plot_volcano <- EnhancedVolcano(\n",
        "    toptable = current_toptable,\n",
        "    lab = current_toptable$SYMBOL,\n",
        "    x = \"logFC\",\n",
        "    y = \"adj.P.Val\",\n",
        "    ylab = bquote(~ -Log[10] ~ \"adjusted pvalue\"),\n",
        "    pCutoff = 0.05,\n",
        "    FCcutoff = 1,\n",
        "    axisLabSize = 12,\n",
        "    captionLabSize = 10,\n",
        "    title = NULL,\n",
        "    subtitle = \"\",\n",
        "    caption = str_c(\"Volcano Plot - \", current_contrast),\n",
        "    pointSize = 0.2,\n",
        "    labSize = 3.0,\n",
        "    legendPosition = \"top\",\n",
        "    legendLabSize = 10,\n",
        "    legendIconSize = 4.0,\n",
        "    colAlpha = 0.5,\n",
        "    gridlines.major = FALSE,\n",
        "    gridlines.minor = FALSE\n",
        ")\n",
        "\n",
        "# Draw plot\n",
        "plot(plot_volcano)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7f6c1793-cc81-4321-ba81-aa06caaaca80",
      "metadata": {},
      "source": [
        "## Table of top DEGs"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "681fca7b-110f-475a-86a2-12d451a05576",
      "metadata": {},
      "source": [
        "# Gene set analysis"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f4679104-a6d7-4211-9a62-f66d2f9407aa",
      "metadata": {},
      "source": [
        "# Advanced exercises\n",
        "The resources provided below are for optional, advanced exercises:\n",
        "- Gene lists for online database analysis."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "d363f708-6660-4416-915c-a7f1cd144ea1",
      "metadata": {},
      "source": [
        "## List of differentially expressed genes\n",
        "Genes with `adjusted p value < 0.0001` and `logFC < -3 or logFC > 3` in the selected comparison are saved to the `deg_{comparison}.txt` file in the root directory.\n",
        "\n",
        "You can right click and press `Download` to save the list to your device.\n",
        "\n",
        "The list can be used at these online services:\n",
        "\n",
        "- [STRING-db Protein-protein interaction analysis](https://www.string-db.org)\n",
        "- [g:Profiler gene set analysis](https://biit.cs.ut.ee/gprofiler/gost)\n",
        "- [DAVID gene set analysis](https://davidbioinformatics.nih.gov/)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "id": "27df7d1f-64ab-482e-809d-c3f4bafbfe9e",
      "metadata": {
        "editable": true,
        "slideshow": {
          "slide_type": ""
        },
        "tags": [],
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#| echo: true\n",
        "#| warning: false\n",
        "#| output: true\n",
        "\n",
        "# Define current contrast\n",
        "current_contrast <- \"ml_vs_basal\"\n",
        "\n",
        "# Get DEG list\n",
        "list_deg <- topTable(\n",
        "    fit = GSE_fitted,\n",
        "    coef = current_contrast,\n",
        "    number = Inf,\n",
        "    sort.by = \"logFC\",\n",
        "    p.value = 0.0001,\n",
        "    lfc = 3\n",
        ")\n",
        "\n",
        "list_deg <- list_deg$SYMBOL |>\n",
        "    na.omit()\n",
        "\n",
        "# Save DEG list\n",
        "writeLines(\n",
        "    list_deg,\n",
        "    con = str_c(\"deg_\", current_contrast, \".txt\", sep = \"\")\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2f7a01a7-384e-421f-aa98-211cc0998062",
      "metadata": {
        "editable": true,
        "slideshow": {
          "slide_type": ""
        },
        "tags": []
      },
      "source": [
        "# Advanced details"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "220b41bd-e8e4-4a0e-a471-568703a4d729",
      "metadata": {},
      "source": [
        "## R syntax - variable assignment\n",
        "\n",
        "In some cases throughout this workshop, you will need to assign one or many genes to a variable.\n",
        "\n",
        "To assign one gene, use: `x <- \"Gene\"`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8c6dcd56-239f-43dc-93d6-0b2c8f0abc08",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Demonstration: assignment of one character to one variable\n",
        "x <- \"Gene\"\n",
        "\n",
        "# Show content of x\n",
        "print(x)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "db9c3faf-da4b-47b4-9236-4c93deb2f947",
      "metadata": {},
      "source": [
        "To assign multiple genes (a vector of gene symbols), use: `x <- c(\"Gene1\", \"Gene2\", \"Gene3\", \"Gene4\")`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "2b6c6da5-9348-4822-b1bd-66a5502f3552",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Demonstration: assignment of a vector to one variable\n",
        "x <- c(\"Gene1\", \"Gene2\", \"Gene3\", \"Gene4\")\n",
        "\n",
        "# Longer vectors can span over multiple lines\n",
        "x <- c(\n",
        "    \"Gene1\",\n",
        "    \"Gene2\",\n",
        "    \"Gene3\",\n",
        "    \"Gene4\"\n",
        ")\n",
        "\n",
        "# Show content of x\n",
        "print(x)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6bc495c4-dbe6-4116-ad52-f61713f84585",
      "metadata": {},
      "source": [
        "Common errors:\n",
        "- **R is case-sensitive!** Make sure you use the correct case when referring to your genes.\n",
        "- Characters must be wrapped in \"quotation marks\"."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6568026b-7922-440a-b0c3-1603fcfb0e1d",
      "metadata": {},
      "source": [
        "## R syntax - exploring objects in this project\n",
        "The uploaded `analysed_GSE63310.Rdata` includes the following objects:\n",
        "- `GSE_logCPM`: Log-counts per million RNA-seq data (with accompanying gene and sample information)\n",
        "- `GSE_fitted`: Test results of RNA-seq data (with accompanying gene and sample information)\n",
        "- `GSE_camera`: Gene set analysis results of RNA-seq data\n",
        "- `GSE_pca`: Principal Component Analysis results of RNA-seq data\n",
        "\n",
        "Below are code examples for accessing these data:"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0169c3f8-af06-4df6-ad7f-56e4d240819f",
      "metadata": {},
      "source": [
        "- Accessing log-counts per million data"
      ]
    },
    {
      "cell_type": "raw",
      "id": "9b76b592-855b-4db0-9eeb-db8474a41232",
      "metadata": {
        "scrolled": true
      },
      "source": [
        "GSE_logCPM[[\"E\"]]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "91067f9c-c52b-4268-b5a6-5fbee7c24c95",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Demonstrate:\n",
        "head(GSE_logCPM[[\"E\"]])"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e0a87297-b24c-4bac-bcf4-d75aec7f1e33",
      "metadata": {},
      "source": [
        "- Accessing gene information"
      ]
    },
    {
      "cell_type": "raw",
      "id": "2a62f556-ac8c-4446-9d7d-262e8be69a90",
      "metadata": {},
      "source": [
        "GSE_logCPM[[\"genes\"]]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e0a171ef-4d20-4f72-a015-d02d9930034c",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Demonstrate:\n",
        "head(GSE_logCPM[[\"genes\"]])"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "89768fa9-ca03-4e32-b0b1-544ba9b6a462",
      "metadata": {},
      "source": [
        "- Accessing sample information"
      ]
    },
    {
      "cell_type": "raw",
      "id": "c48e5090-c887-4f95-b8ec-6675782980cf",
      "metadata": {},
      "source": [
        "GSE_logCPM[[\"targets\"]]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "be4e4dc1-4a8e-4dc8-99fc-237a20d407ed",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Demonstrate\n",
        "head(GSE_logCPM[[\"targets\"]])"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e34a73e3-6449-4097-8c36-1fc077ef9303",
      "metadata": {},
      "source": [
        "- Accessing comparisons definition (in this context, comparisons are called \"contrasts\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d48812ae-a958-4135-bd21-06de18f4799d",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Accessing contrasts\n",
        "print(GSE_fitted[[\"contrasts\"]])\n",
        "\n",
        "# Accessing names of contrasts\n",
        "print(colnames(GSE_fitted[[\"contrasts\"]]))"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "30296f1b-d202-490d-9edd-2c826d111fa4",
      "metadata": {},
      "source": [
        "- Accessing gene set analysis results"
      ]
    },
    {
      "cell_type": "raw",
      "id": "8396a01c-1323-4314-88d6-aff652e0101f",
      "metadata": {
        "collapsed": true,
        "editable": true,
        "jupyter": {
          "outputs_hidden": true
        },
        "scrolled": true,
        "slideshow": {
          "slide_type": ""
        },
        "tags": []
      },
      "source": [
        "# Note: use GSE_camera[[\"name of contrasts\"]]\n",
        "GSE_camera[[\"ml_vs_lp\"]]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f7a122b4-44d5-4c57-8634-746243df9ccb",
      "metadata": {
        "trusted": true,
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Demonstrate\n",
        "head(GSE_camera[[\"ml_vs_lp\"]])"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "R (webR)",
      "language": "R",
      "name": "webR"
    },
    "language_info": {
      "file_extension": ".R",
      "mimetype": "text/x-rsrc",
      "name": "R",
      "version": "4.5.1"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
